src/cursor.c

```diff
@@ -891,15 +935,24 @@ cursor_axis(struct wl_listener *listener, void *data)
 	 */
 	struct seat *seat = wl_container_of(listener, seat, cursor_axis);
 	struct wlr_pointer_axis_event *event = data;
+	struct server *server = seat->server;
+	struct cursor_context ctx = get_cursor_context(server);
 	wlr_idle_notify_activity(seat->wlr_idle, seat->seat);
 
-	/* Make sure we are sending the events to the surface under the cursor */
-	cursor_update_focus(seat->server);
+	/* Bindings swallow mouse events if activated */
+	bool triggered_axis_binding = handle_cursor_axis(server, &ctx, event);
+
+	if (ctx.surface && !triggered_axis_binding) {
+		/* Notify client of axis event */
 
-	/* Notify the client with pointer focus of the axis event. */
-	wlr_seat_pointer_notify_axis(seat->seat, event->time_msec,
-		event->orientation, event->delta, event->delta_discrete,
-		event->source);
+		/* Make sure we are sending the events to the surface under the cursor */
+		cursor_update_focus(server);
```

Comment:
We could likely use the `cursor_update_common()` variant here instead as we already have the cursor context and thus don't have to do the hitbox detection again.